src/Components/Web.JS/src/Rendering/FocusOnNavigate.ts

```diff
@@ -0,0 +1,99 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+import { domFunctions } from '../DomWrapper';
+import { EnhancedNavigationStartEvent, JSEventRegistry } from '../Services/JSEventRegistry';
+
+const customElementName = 'blazor-focus-on-navigate';
+let currentFocusOnNavigateElement: FocusOnNavigateElement | null = null;
+let allowFocusOnEnhancedLoad = false;
+
+export function enableFocusOnNavigate(jsEventRegistry: JSEventRegistry) {
+  customElements.define(customElementName, FocusOnNavigateElement);
+  jsEventRegistry.addEventListener('enhancednavigationstart', onEnhancedNavigationStart);
+  jsEventRegistry.addEventListener('enhancedload', onEnhancedLoad);
+  document.addEventListener('focusin', onFocusIn);
+  document.addEventListener('focusout', onFocusOut);
+
+  // Focus the element on the initial page load.
+  if (document.readyState === 'loading') {
+    // There may be streaming updates on the initial page load, so we want to allow
+    // those to add elements matching the active selector.
+    allowFocusOnEnhancedLoad = true;
+    document.addEventListener('DOMContentLoaded', afterInitialPageLoad);
+  } else {
+    afterInitialPageLoad();
+  }
+}
+
+function afterInitialPageLoad() {
+  tryApplyFocus(/* forceMoveFocus */ false);
+}
+
+function onEnhancedNavigationStart(ev: EnhancedNavigationStartEvent) {
+  // Only focus on enhanced load if the enhanced navigation is not a form post.
+  allowFocusOnEnhancedLoad = ev.options.method !== 'post';
```

Comment:
This rule could use some clarification in the comment. My guess is that you're trying to differentiate "navigating to a new place" (should change focus) vs "updating the existing page" (should not).

There's a slightly related rule in `NavigationEnhancement.ts` where we check `isForSamePath` to determine whether we think the new URL is conceptually equivalent to the old one - it compares the URL up to, but excluding, the querystring. Arguably there would be some benefit to following the same rule here since that would give us fewer new rules to explain/justify, and is arguably more general than doing it on HTTP method alone. There are certainly cases with an enhanced GET request where the UI is conceptually just updating and not going to a new page, e.g., pagination links on a grid, just as there are cases where a POST conceptually goes to a new page.

But altogether I guess this is a subjective call to make.  I'm fine with whatever subjective choice you prefer here. If you could slightly expand the code comment to specify the rationale for whatever rule you prefer, I think that's totally sufficient.

Reply:
I think is better if we are consistent with what we do in `NavigationEnhancement.ts`

Reply:
This has been updated to reuse the same logic in `NavigationEnhancement.ts`